### PR TITLE
os: TempDir(): obey TMPDIR on unix, TMP on win, etc

### PR DESCRIPTION
**Pull request by dkegel-fastly**
Uses upstream go code, lightly adjusted.

Adds smoke test.

### CODE REVIEW

**Comment by niaow** [comment]
If this contains upstream Go code, appropriate license headers should be added.

**Review by dgryski — APPROVED** [review]
LGTM

**Comment by deadprogram** [comment]
@dkegel-fastly this PR now needs to be rebased against `dev` to resolve merge conflicts, if you please.

File: src/os/file_anyos_test.go [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+)
+
+// TestTempDir assumes that the test environment has a filesystem with a working temp dir.
+func TestTempDir(t *testing.T) {
+	name := TempDir() + "/_os_test_TestTempDir"
+	Remove(name)
+	f, err := OpenFile(name, O_RDWR|O_CREATE, 0644)
+	if err != nil {
+		t.Errorf("OpenFile %s: %s", name, err)
+		return
```
Comment (aykevl) [inline src/os/file_anyos_test.go:15]: Suggestion: I believe this can be replaced with `t.Fatal`.

**Review by aykevl — APPROVED** [review]
Looks reasonable to me!

File: src/os/file_anyos_test.go (hunk 1 above)
Comment (dkegel-fastly) [reply src/os/file_anyos_test.go:15]: It would still need the return at the moment, I went for lower cognitive dissonance :-)

**Comment by deadprogram** [comment]
Thanks everyone! Now merging.

File: src/os/file_anyos_test.go (hunk 1 above)
Comment (aykevl) [reply src/os/file_anyos_test.go:15]: Fair enough. Indeed, `t.Fatal` isn't obviously returning (even though the effect is similar).

**Pull request merged**